### PR TITLE
[Issue #6] Optimizer toggle to filter out equipped relics

### PR DESCRIPTION
**Pull request by cnojima**
New checkbox control in Optimizer to dis/enable consideration of relics currently equipped by characters.

Screenshot 2024-01-19 at 10 57 16 PM

### CODE REVIEW

File: src/components/OptimizerForm.js [hunk 1]
```diff
@@ -214,11 +214,6 @@ export default function OptimizerForm() {
     }
   }, [selectedLightCone])
 
-  const onChangeStatDisplay = ({ target: { value } }) => {
```
Comment (cnojima) [inline src/components/OptimizerForm.js:217]: refactored to `components/optimizerTab/OptimizerOptions.js`

File: src/components/OptimizerForm.js [hunk 2]
```diff
@@ -711,124 +708,7 @@ export default function OptimizerForm() {
               </Flex>
             </FormCard>
 
-            <FormCard>
```
Comment (cnojima) [inline src/components/OptimizerForm.js:714]: refactored to `components/optimizerTab/OptimizerOptions.js`

File: src/components/optimizerTab/OptimizerOptions.js [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+            checkedChildren={<CheckOutlined />}
+            unCheckedChildren={<CloseOutlined />}
+            defaultChecked
+            style={{ width: 45, marginRight: 10 }}
+          />
+        </Form.Item>
+        <Text>Keep current relics</Text>
+      </Flex>
+
+      <Flex align='center'>
```
Comment (cnojima) [inline src/components/optimizerTab/OptimizerOptions.js:65]: net-new control here.

File: src/lib/db.js [hunk 4]
```diff
@@ -100,6 +100,7 @@ window.store = create((set) => ({
     subStats: [],
   },
   setRelicTabFilters: (x) => set(() => ({ relicTabFilters: x })),
+  setFilteredRelics: (relics) => set(() => ({ filteredRelics: relics })),
```
Comment (cnojima) [inline src/lib/db.js:103]: expose filteredRelics (all filters applied) via hook

File: src/lib/db.js [hunk 5]
```diff
@@ -164,6 +165,10 @@ export const DB = {
     relicsById[relic.id] = relic
     global.store.getState().setRelicsById(relicsById)
   },
+  setFilteredRelics: (relics) => {
```
Comment (cnojima) [inline src/lib/db.js:168]: expose setFilteredRelics by global store method

File: src/lib/db.js [hunk 6]
```diff
@@ -164,6 +165,10 @@ export const DB = {
...
+    global.store.getState().setFilteredRelics(relics)
+  },
+  getFilteredRelics: () => global.store.getState().filteredRelics,
```
Comment (cnojima) [inline src/lib/db.js:171]: hook `getFilteredRelics`

File: src/lib/hint.js [hunk 7]
```diff
@@ -125,17 +125,18 @@ export const Hint = {
...
-          <p>Keep current relics - The character must use its currently equipped items, and the optimizer will try to fill in empty slots</p>
-          <p>Enhance / grade - Select the minimum enhance to search for and minimum stars for relics to include</p>
-          <p>Stat display - Select which format of stats to apply to filters and display in the table. Base stats are the values you would see in the ingame character menu. Combat stats take into account all the buffs and passives applied to the character in combat as they perform an attack.</p>
+          <p><strong>Rank filter</strong> - Rank characters by dragging them on the character page, and when enabled, characters may only take relics from lower ranked characters</p>
+          <p><strong>Maxed main stat</strong> - Assume the main stat for relics are maxed</p>
+          <p><strong>Keep current relics</strong> - The character must use its currently equipped items, and the optimizer will try to fill in empty slots</p>
+          <p><strong>Use equipped</strong> - If OFF, Optimizer will NOT consider relics currently equipped by a character</p>
```
Comment (cnojima) [inline src/lib/hint.js:137]: updated tooltip.  other changes are to bold control name vs. the description

File: src/lib/optimizer.js [hunk 8]
```diff
@@ -111,7 +111,7 @@ export const Optimizer = {
       }
     }
 
-    let relics = Utils.clone(DB.getRelics());
+    let relics = Utils.clone(DB.getFilteredRelics());
```
Comment (cnojima) [inline src/lib/optimizer.js:114]: @fribbels DB.getRelics() seems to grab the full set here.
e.g., Permutations count would show "250,000", but optimizing would result in millions of hits (using my personal kel-z import data).

had to give explicit get/set to get optimizer to correctly calc with filtered out relics.  maybe a bug from before?  or did i miss the schema/filter logic completely somewhere else?

File: src/lib/relicFilters.js [hunk 9]
```diff
@@ -85,14 +85,28 @@ export const RelicFilters = {
...
-  
+
     return out;
   },
 
   applyEnhanceFilter: (request, relics) => {
     return relics.filter(x => x.enhance >= request.enhance);
   },
 
+  applyEquippedFilter: (request, relics) => {
```
Comment (cnojima) [inline src/lib/relicFilters.js:96]: actual filter logic here.  DB/store relics set from import is missing foreign key to character.equipped.   e.g., all the relics "equippedBy" IDs are `undefined`.

i had to manually create the blacklist here by iterating over all character.equipped maps.

maybe a backlog bug to repair metadata to ensure links are maintained during import?

File: src/lib/optimizer.js (hunk 8 above)
Comment (fribbels) [reply src/lib/optimizer.js:114]: @cnojima 

This was duplicated code that I've been meaning to refactor - but I think the issue this introduces is that if you set filters -> calculate permutations -> save filtered relics -> but then edit a relic / import new relics -> the optimizer uses the cached version of the old relics not the updated ones. 

The filter methods in this file will get cleaned up into RelicFilters so we don't have this duplication and then maybe we can just called RelicFilters.applyAllFilters() in one place.

For now I think we can add a applyEquippedFilter to this duplicated filter list to resolve your 250,000 vs 2m discrepancy

Comment (fribbels) [reply src/lib/optimizer.js:114]: https://github.com/fribbels/hsr-optimizer/issues/55

File: src/lib/relicFilters.js (hunk 9 above)
Comment (fribbels) [reply src/lib/relicFilters.js:96]: https://github.com/fribbels/hsr-optimizer/issues/56

**Comment by fribbels** [comment]
Tested and the NaNs are gone, thanks!

**Pull request merged**